resources/type-system/flow-analysis.md

```diff
@@ -0,0 +1,713 @@
+    - `r3` is `push(r0, top(rb) && top(rf))`
+    - `b` is true if `v` is in `N` and otherwise false
+    - `VI3` is the map which maps each variable `v` in the domain of `VIB` and
+      `VIF` to `restrictV(VIB(v), VIF(v), b)`.
+
+- `unreachable(M)` represents the model corresponding to a program location
+  which is unreachable, but is otherwise modeled by flow model `M = FlowModel(r,
+  VI)`, and is defined as `FlowModel(push(pop(r), false), VI)`
+
+- `demoteVariables(M, S)` represents the flow model derived from `M` in which
+  all variables in `S` have been demoted to their declared types.  It is defined
+  as `FlowModel(r, VI1)` where `M` is `FlowModel(r, VI0)` and `VI1` is the map
+  such that:
+    - `VI0` maps `v` to `VM0 = VariableModel(d0, p0, s0, a0, u0, c0)`
+    - If `S` contains `v` then `VI1` maps `v` to `VariableModel(d0, [], s0, a0, u0, c0)`
+    - Otherwise `VI1` maps `v` to `VM0`
+
+
+### Promotion
+
+Promotion policy is defined by the following operations on flow models.
+
+We say that the **current type** of a variable `x` in variable model `VM` is `S` where:
+  - `VM = VariableModel(declared, promoted, tested, assigned, unassigned, captured)`
+  - `promoted = S::l` or (`promoted = []` and `declared = S`)
+
+Policy:
+  - We say that at type `T` is a type of interest for a variable `x` in a set of
+    tested types `tested` if `tested` contains a type `S` such that `T` is `S`,
+    or `T` is **NonNull(`S`)**.
+
+  - We say that a variable `x` is promotable via type test with type `T` given
+    variable model `VM` if
+    - `VM = VariableModel(declared, promoted, tested, assigned, unassigned, captured)`
+    - and `captured` is false
+    - and `S` is the current type of `x` in `VM`
+    - and not `S <: T`
+    - and `T <: S` or (`S` is `X extends R` and `T <: R`) or (`S` is `X & R` and
+      `T <: R`)
+
+  - We say that a variable `x` is promotable via initialization given variable
+    model `VM` if:
+    - `VM = VariableModel(declared, promoted, tested, assigned, unassigned, captured)`
+    - and `captured` is false
+    - and `promoted` is empty
+    - and `x` is declared with no explicit type and no initializer
+    - and `assigned` is false and `unassigned` is true
+
+  - We say that a variable `x` is promotable via assignment of an expression of
+    type `T` given variable model `VM` if
+    - `VM = VariableModel(declared, promoted, tested, assigned, unassigned, captured)`
+    - and `captured` is false
+    - and `S` is the current type of `x` in `VM`
+    - and `T <: S` and not `S <: T`
+    - and `T` is a type of interest for `x` in `tested`
+
+  - We say that a variable `x` is demotable via assignment of an expression of
+    type `T` given variable model `VM` if
+    - `VM = VariableModel(declared, promoted, tested, assigned, unassigned, captured)`
+    - and `captured` is false
+    - and declared::promoted contains a type `S` such that `T` is `S` or `T` is
+      **NonNull(`S`)**.
```

Comment:
The "or `T` is **NonNull(`S`)**" part of this isn't implemented either.  Also tracking in https://github.com/dart-lang/sdk/issues/40011